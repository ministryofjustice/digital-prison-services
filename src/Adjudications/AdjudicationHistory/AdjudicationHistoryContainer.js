import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import axios from 'axios'
import moment from 'moment'
import Page from '../../Components/Page'
import AdjudicationHistoryForm from './AdjudicationHistoryForm'
import AdjudicationHistoryTable from './AdjudicationHistoryTable'
import {
  setAdjudicationHistoryResults,
  setAdjudicationHistoryFilter,
  setAdjudicationHistoryPageNumber,
  setAdjudicationHistoryPageSize,
} from '../../redux/actions'

export class AdjudicationHistoryContainer extends Component {
  componentDidMount = () => {
    this.loadAdjudications({})
  }

  componentDidUpdate = prevProps => {
    const { offenderNumber } = this.props

    if (offenderNumber !== prevProps.offenderNumber) {
      this.loadAdjudications({})
    }
  }

  changePage = async page => {
    const { setAdjudicationHistoryPageNumber: setPageNumber, fieldValues } = this.props
    await setPageNumber(page)
    this.updateResults(fieldValues)
  }

  changePerPage = async pageSize => {
    const { setAdjudicationHistoryPageSize: setPageSize } = this.props
    await setPageSize(Number(pageSize))
    this.changePage(0)
  }

  applyFilter = async formValues => {
    const { setAdjudicationHistoryPageNumber: setPageNumber } = this.props
    await setPageNumber(0)
    this.updateResults({ ...formValues })
  }

  reset = () => this.updateResults({ offenceId: null, establishment: null, fromDate: null, toDate: null })

  updateResults = fields => {
    const { setAdjudicationHistoryFilter: setFilter } = this.props
    setFilter(fields)
    return this.loadAdjudications(fields)
  }

  loadAdjudications = async ({ offenceId, establishment, fromDate, toDate }) => {
    const {
      setAdjudicationHistoryResults: setResults,
      offenderNumber,
      pageNumber,
      pageSize,
      setLoadedDispatch,
      resetErrorDispatch,
      handleError,
    } = this.props

    resetErrorDispatch()
    setLoadedDispatch(false)
    try {
      const { data: adjudications, headers } = await axios.get(`/api/adjudications/${offenderNumber}`, {
        params: {
          offenceId,
          agencyId: establishment,
          fromDate: fromDate && fromDate.format('YYYY-MM-DD'),
          toDate: toDate && toDate.format('YYYY-MM-DD'),
        },
        headers: {
          'Page-Offset': pageNumber * pageSize,
          'Page-Limit': pageSize,
        },
      })

      const totalRecords = parseInt(headers['total-records'], 10)

      setResults({
        ...adjudications,
        totalRecords,
      })
    } catch (error) {
      handleError(error)
    }
    setLoadedDispatch(true)
  }

  render = () => (
    <Page title="Adjudication history">
      <AdjudicationHistoryForm search={this.applyFilter} reset={this.reset} />
      <AdjudicationHistoryTable changePage={this.changePage} changePerPage={this.changePerPage} />
    </Page>
  )
}

AdjudicationHistoryContainer.propTypes = {
  // props
  offenderNumber: PropTypes.string.isRequired,
  handleError: PropTypes.func.isRequired,
  setLoadedDispatch: PropTypes.func.isRequired,
  resetErrorDispatch: PropTypes.func.isRequired,

  // mapStateToProps
  pageNumber: PropTypes.number.isRequired,
  pageSize: PropTypes.number.isRequired,
  setAdjudicationHistoryResults: PropTypes.func.isRequired,
  setAdjudicationHistoryFilter: PropTypes.func.isRequired,
  setAdjudicationHistoryPageNumber: PropTypes.func.isRequired,
  setAdjudicationHistoryPageSize: PropTypes.func.isRequired,
  fieldValues: PropTypes.shape({
    establishment: PropTypes.string,
    offenceId: PropTypes.string,
    fromDate: PropTypes.instanceOf(moment),
    toDate: PropTypes.instanceOf(moment),
  }).isRequired,
}

const mapStateToProps = state => ({
  pageNumber: state.adjudicationHistory.pageNumber,
  pageSize: state.adjudicationHistory.pageSize,
  fieldValues: {
    establishment: state.adjudicationHistory.establishment,
    offenceId: state.adjudicationHistory.offenceId,
    fromDate: state.adjudicationHistory.fromDate,
    toDate: state.adjudicationHistory.toDate,
  },
})

const mapDispatchToProps = {
  setAdjudicationHistoryResults,
  setAdjudicationHistoryFilter,
  setAdjudicationHistoryPageNumber,
  setAdjudicationHistoryPageSize,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AdjudicationHistoryContainer)
