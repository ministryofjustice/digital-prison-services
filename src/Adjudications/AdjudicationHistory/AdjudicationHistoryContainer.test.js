import React from 'react'
import { shallow } from 'enzyme'
import moment from 'moment'

import AdjudicationHistoryContainer from './AdjudicationHistoryContainer'

const initialState = {
  adjudicationHistory: {
    pageNumber: 0,
    pageSize: 20,
  },
}

describe('Adjudication History container', () => {
  const now = moment('2019-10-10T21:00:00Z')
  const store = {}
  const history = {}
  beforeEach(() => {
    store.getState = jest.fn()
    store.subscribe = jest.fn()
    store.dispatch = jest.fn()
    history.push = jest.fn()
    history.replace = jest.fn()
    store.getState.mockReturnValue(initialState)
  })

  it('should render correctly', () => {
    const wrapper = shallow(
      <AdjudicationHistoryContainer
        now={now}
        store={store}
        offenderNumber="AA123BB"
        handleError={jest.fn()}
        setLoadedDispatch={jest.fn()}
        resetErrorDispatch={jest.fn()}
        history={history}
      />
    )

    expect(wrapper.dive()).toMatchSnapshot()
  })
})
