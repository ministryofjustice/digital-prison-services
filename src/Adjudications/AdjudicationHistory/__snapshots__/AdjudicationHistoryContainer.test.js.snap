// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adjudication History container should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AdjudicationHistoryContainer
    fieldValues={
      Object {
        "establishment": undefined,
        "fromDate": undefined,
        "offenceId": undefined,
        "toDate": undefined,
      }
    }
    handleError={[MockFunction]}
    history={
      Object {
        "push": [MockFunction],
        "replace": [MockFunction],
      }
    }
    now={"2019-10-10T21:00:00.000Z"}
    offenderNumber="AA123BB"
    pageNumber={0}
    pageSize={20}
    resetErrorDispatch={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    setAdjudicationHistoryFilter={[Function]}
    setAdjudicationHistoryPageNumber={[Function]}
    setAdjudicationHistoryPageSize={[Function]}
    setAdjudicationHistoryResults={[Function]}
    setLoadedDispatch={
      [MockFunction] {
        "calls": Array [
          Array [
            false,
          ],
        ],
      }
    }
    store={
      Object {
        "dispatch": [MockFunction],
        "getState": [MockFunction] {
          "calls": Array [
            Array [],
            Array [],
          ],
        },
        "subscribe": [MockFunction] {
          "calls": Array [
            Array [
              [Function],
            ],
          ],
        },
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [MockFunction],
          "getState": [MockFunction] {
            "calls": Array [
              Array [],
              Array [],
            ],
          },
          "subscribe": [MockFunction] {
            "calls": Array [
              Array [
                [Function],
              ],
            ],
          },
        },
        "unsubscribe": undefined,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Connect(AdjudicationHistoryForm)
          reset={[Function]}
          search={[Function]}
        />,
        <Connect(AdjudicationHistoryTable)
          changePage={[Function]}
          changePerPage={[Function]}
        />,
      ],
      "title": "Adjudication history",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "reset": [Function],
          "search": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "changePage": [Function],
          "changePerPage": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Connect(AdjudicationHistoryForm)
            reset={[Function]}
            search={[Function]}
          />,
          <Connect(AdjudicationHistoryTable)
            changePage={[Function]}
            changePerPage={[Function]}
          />,
        ],
        "title": "Adjudication history",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "reset": [Function],
            "search": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "changePage": [Function],
            "changePerPage": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
